cpp/src/arrow/dataset/scanner.h

```diff
@@ -152,6 +152,9 @@ class ARROW_DS_EXPORT ScanTask {
   /// resulting from the Scan. Execution semantics are encapsulated in the
   /// particular ScanTask implementation
   virtual Result<RecordBatchIterator> Execute() = 0;
```

Comment:
So to make sure, SafeExecute/SafeVisit are just a way to safely integrate async readers while we still have SyncScanner?

Reply:
(Note #10062 removes the ScanTask bindings from Python, leaving just CGlib.)

Reply:
Correct.  These methods can go away when `SyncScanner` goes away.  Removing the `ScanTask` bindings won't be enough (although I'm happy to hear they are removed) because `SyncScanner::ToTable` and the synchronous dataset write both rely on scanning being safe to call (ala nested deadlock) from a CPU executor thread.

Reply:
Also, this will probably only be needed for CSV.  With parquet & IPC you created the async version in parallel with the sync verison.  With CSV there is just one (async) code path.